torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -2409,6 +2428,19 @@ uint64_t ProcessGroupNCCL::WorkNCCL::getSequencenumber() const {
   return seq_;
 }
 
+void ProcessGroupNCCL::assignTimeoutToWork(
+    c10::intrusive_ptr<ProcessGroupNCCL::WorkNCCL> work,
+    c10::intrusive_ptr<ProcessGroupNCCL::Options> option) {
+  std::chrono::milliseconds timeout = option->timeout;
+  std::unique_lock<std::mutex> lock(mtxTimeoutExtension_);
```

Comment:
What's locking discipline on a PG overall? If I mutably access it from multiple threads is it expected that I implicitly have taken out another lock, or is PG responsible for locking? Latter is pretty unusual for PyTorch C++ API design

Reply:
I don't think I am following the question here, are you asking whether the lock is implicitly acquired or explicitly lock/unlock? If you are asking about this, I am afraid PG need to be responsible for the lock...

Reply:
So, let's say you design a struct in C++ with some methods, and this struct is potentially accessible from multiple threads. You have to decide what locking strategy is. One approach is to say mutations from multiple threads not allowed, you have to make sure you never race mutations. at::Tensor works this way; multithread access OK, but not multithread writes. Another approach is to modify every method to implicitly take out a lock when you call it, so method calls are thread safe without any special synchronization. This sort of thing is very popular in Java because they have a dedicated keyword synchronized that does it.

Reply:
I see. Thanks for the detailed explanation. I think in PGNCCL we unfortunately don't do single thread write (aka the at::Tensor way you mentioned here) for example, we let the main thread to add the Work into queue and then erase it in the watchdog thread. For this PR specifically, we also do the similar thing, for example in `ephemeralTimeoutSet_` we set it in the main thread but reset in the watchdog thread. I guess the reason here is because we use watchdog thread as a side thread to do all cleanups and completion check. And the mutex we add here is like what you mentioned here as a implicit lock. (watchdog only calls on function, lol)